### PR TITLE
Torquefoccurrent simulation

### PR DESCRIPTION
**Pull request by Atman-Singh**

### CODE REVIEW

**Comment by HarryXChen3** [comment]
Closes #23

File: src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java [hunk 1]
```diff
@@ -38,9 +38,12 @@ public class SwerveModuleIOTalonFXSim implements SwerveModuleIO {
     private final TalonFXConfiguration driveTalonFXConfiguration = new TalonFXConfiguration();
     private final TalonFXConfiguration turnTalonFXConfiguration = new TalonFXConfiguration();
 
-    private final VelocityVoltage velocityVoltage;
+    private final VelocityTorqueCurrentFOC velocityTorqueCurrentFOC;
+
```
Comment (HarryXChen3) [inline src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java:42]: Is there a reason why these fields are so spaced out from each other with empty newlines?

File: src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java [hunk 2]
```diff
@@ -93,8 +96,11 @@ public SwerveModuleIOTalonFXSim(
         );
         this.turnSim.attachFeedbackSensor(new SimPhoenix6CANCoder(turnEncoder));
 
-        this.velocityVoltage = new VelocityVoltage(0);
+        this.velocityTorqueCurrentFOC = new VelocityTorqueCurrentFOC(0);
         this.positionVoltage = new PositionVoltage(0);
+        this.odometryThreadRunner = odometryThreadRunner;
+        this.odometryThreadRunner.registerControlRequest(driveMotor, velocityTorqueCurrentFOC, driveMotor::setControl);
+        this.odometryThreadRunner.registerControlRequest(turnMotor, positionVoltage, turnMotor::setControl);
```
Comment (HarryXChen3) [inline src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java:103]: Were these calls were missing from the previous simulation implementation?

File: src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java [hunk 3]
```diff
@@ -127,29 +133,42 @@ public void config() {
...
         turnTalonFXConfiguration.Feedback.RotorToSensorRatio = Constants.Swerve.Modules.TURNER_GEAR_RATIO;
-        turnTalonFXConfiguration.ClosedLoopGeneral.ContinuousWrap = true;
-        turnTalonFXConfiguration.MotorOutput.NeutralMode = NeutralModeValue.Coast;
+        turnTalonFXConfiguration.MotorOutput.NeutralMode = NeutralModeValue.Brake;
         turnTalonFXConfiguration.MotorOutput.Inverted = turnInvertedValue;
+
+        turnTalonFXConfiguration.MotionMagic.MotionMagicCruiseVelocity =
+                100.0 / Constants.Swerve.Modules.TURNER_GEAR_RATIO;
+        turnTalonFXConfiguration.MotionMagic.MotionMagicExpo_kV = 0.12 * Constants.Swerve.Modules.TURNER_GEAR_RATIO;
+        turnTalonFXConfiguration.MotionMagic.MotionMagicExpo_kA = 0.1;
```
Comment (HarryXChen3) [inline src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java:164]: `MotionMagic` configurations should be removed when not using motion magic control modes

**Review by HarryXChen3 — CHANGES_REQUESTED** [review]
Did we check that these changes still work properly in simulation? We don't have unit/integration tests to check this yet so we should manually verify that the simulation still behaves as expected.

File: src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java (hunk 2 above)
Comment (Atman-Singh) [reply src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java:103]: yes

File: src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java (hunk 1 above)
Comment (MqxS) [reply src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java]: Fixed.

File: src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java (hunk 3 above)
Comment (MqxS) [reply src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java]: Removed.

**Review by MqxS — APPROVED** [review]
lgtm

**Review by MqxS — DISMISSED** [review]
fixed.

File: src/main/java/frc/robot/subsystems/drive/OdometryThreadRunner.java [hunk 4]
```diff
@@ -122,7 +122,7 @@ public Class<State> getTypeClass() {
 
             @Override
             public String getTypeString() {
-                return "struct:Swerve.OdometryThreadRunner.State";
+                return "struct:OdometryThreadRunner.State";
```
Comment (HarryXChen3) [inline src/main/java/frc/robot/subsystems/drive/OdometryThreadRunner.java:125]: This fix is unrelated to this PR and should be included with it - it should get its own branch and have it PR'ed separately.

File: src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java [hunk 5]
```diff
@@ -127,29 +132,35 @@ public void config() {
         //  which means that TorqueCurrent.PeakForwardTorqueCurrent and related won't affect it
         final InvertedValue driveInvertedValue = InvertedValue.CounterClockwise_Positive;
         driveTalonFXConfiguration.Slot0 = new Slot0Configs()
-                .withKV(0.973);
-        driveTalonFXConfiguration.TorqueCurrent.PeakForwardTorqueCurrent = 60;
-        driveTalonFXConfiguration.TorqueCurrent.PeakReverseTorqueCurrent = -60;
+                .withKP(50)
+                .withKS(4.796)
+                .withKA(2.549);
```
Comment (HarryXChen3) [inline src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java:137]: So are these FF and PID gains appropriate for simulation? If they haven't been tuned specifically for simulation, they should be, as currently, the system dynamics differ between the simulated swerve and the real robot swerve.

File: src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java [hunk 6]
```diff
@@ -127,29 +132,35 @@ public void config() {
...
         driveTalonFXConfiguration.MotorOutput.Inverted = driveInvertedValue;
         driveMotor.getConfigurator().apply(driveTalonFXConfiguration);
 
         final InvertedValue turnInvertedValue = InvertedValue.Clockwise_Positive;
         turnTalonFXConfiguration.Slot0 = new Slot0Configs()
-                .withKP(40);
-        turnTalonFXConfiguration.Voltage.PeakForwardVoltage = 6;
-        turnTalonFXConfiguration.Voltage.PeakReverseVoltage = -6;
+                .withKP(30)
+                .withKS(0.5);
```
Comment (HarryXChen3) [inline src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java:151]: Ditto.

**Review by HarryXChen3 — CHANGES_REQUESTED** [review]
So does the swerve run properly in simulation with no regressions compared to using `VelocityVoltage`? I left a few comments that need to be addressed.

File: src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java (hunk 6 above)
Comment (MqxS) [reply src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java:151]: These are from master. They seem to work.

File: src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java (hunk 5 above)
Comment (MqxS) [reply src/main/java/frc/robot/subsystems/drive/SwerveModuleIOTalonFXSim.java:137]: These are from master. They seem to work.

File: src/main/java/frc/robot/subsystems/drive/OdometryThreadRunner.java (hunk 4 above)
Comment (MqxS) [reply src/main/java/frc/robot/subsystems/drive/OdometryThreadRunner.java]: Added back.

**Review by HarryXChen3 — APPROVED** [review]
Ok, LGTM!
